openedx/stanford/common/djangoapps/timed_problem/models.py

```diff
@@ -0,0 +1,112 @@
+import datetime
+
+from django.utils.timezone import UTC
+
+from xblock.fields import Integer
+from xblock.fields import Scope
+
+from xmodule.fields import Date
+
+_ = lambda text: text
+
+
+class WarningFieldSingleValueMixin(object):
+    """
+    Wraps around the actual value of another xmodule.field.Field.
+    If the value of this field is set on the front-end, a warning
+    from the backend is shown
+    """
+    def __init__(self, warning=None, warning_condition=None, **kwargs):
+        super(WarningFieldSingleValueMixin, self).__init__(**kwargs)
+        self._values = {
+            "warning": warning,
+        }
+
+
+class IntegerWithWarningField(WarningFieldSingleValueMixin, Integer):
+    """
+    Integer with a warning field built-in. The warning is triggered
+    when the field name is specified under a certain condition
+    """
+    def __init__(self, **kwargs):
+        super(IntegerWithWarningField, self).__init__(**kwargs)
+
+
+class TimedCapaFields(object):
+    minutes_before_warning = Integer(
+        help=_(
+            'Number of minutes at which the student will be issued a time warning. '
+            'Works only when minutes_allowed is non-zero'
+        ),
+        default=1,
+        scope=Scope.settings,
+    )
+    minutes_allowed = IntegerWithWarningField(
+        display_name=('Minutes Allowed'),
+        help=_(
+            'Number of minutes allowed to finish this assessment. '
+            'Set 0 for no time-limit. '
+            'If there is a time-limit, the student will only be given one attempt.'
+        ),
+        warning=_(
+            'Setting minutes allowed means that this question can only have one attempt, '
+            'regardless of the value of the maximum attempts field.'
+        ),
+        default=0,
+        scope=Scope.settings,
+    )
+    time_started = Date(
+        help=_('time student started this assessment'),
+        scope=Scope.user_state,
+    )
+
+
+class TimedCapaMixin(TimedCapaFields):
```

Comment:
Follow the same inheritance pattern as `CapaMixin` and `CapaFields`.
The naming was intentionally chosen so that `grep`ing for  `CapaMixin` and `CapaFields` will match our timed mixins too :)